tools/odbc/odbc_fetch.cpp

```diff
@@ -12,13 +13,185 @@ using duckdb::OdbcHandleStmt;
+		return SQL_ERROR;
+	}
+
+	return SetCurrentChunk(stmt);
+}
+
+SQLRETURN OdbcFetch::SetCurrentChunk(OdbcHandleStmt *stmt) {
+	if (chunks.empty()) {
+		return SQL_NO_DATA;
+	}
+	// get always the last chunk and set as the current one
+	current_chunk = chunks.back().get();
+	chunk_row = prior_chunk_row = -1;
+
+	return SQL_SUCCESS;
+}
+
+SQLRETURN OdbcFetch::SetPriorCurrentChunk(OdbcHandleStmt *stmt) {
+	prior_chunk_row -= rowset_size;
+
+	if (stack_prior_chunks.empty() && prior_chunk_row < -1) {
+		chunk_row = -1;
+		return SQL_NO_DATA;
+	}
+
+	if (prior_chunk_row < -1) {
+		current_chunk = stack_prior_chunks.top();
+		stack_prior_chunks.pop();
+		prior_chunk_row = current_chunk->size() - rowset_size - 1;
+		if (prior_chunk_row < 0) {
+			prior_chunk_row = -1;
+		}
+	}
+
+	chunk_row = prior_chunk_row;
+
+	return SQL_SUCCESS;
+}
+
+SQLRETURN OdbcFetch::FetchFromBOF() {
+	current_chunk = chunks.front().get();
+	chunk_row = prior_chunk_row = -1;
+	return SQL_SUCCESS;
+}
+
+SQLRETURN OdbcFetch::SetAbsoluteCurrentChunk(OdbcHandleStmt *stmt, SQLLEN fetch_offset) {
+	// there is no current_chunk, it requires fetch next
+	if (!current_chunk || chunks.empty()) {
+		FetchNext(stmt);
+	}
+	// it has reachted the last row
+	if (fetch_offset > (SQLLEN)current_chunk->size() && resultset_end) {
+		return SQL_NO_DATA;
+	}
+
+	// FetchOffset < 0 AND | FetchOffset | <= LastResultRow
+	if ((fetch_offset < 0) && ((duckdb::idx_t)std::abs(fetch_offset) <= current_chunk->size())) {
+		// LastResultRow + FetchOffset + 1
+		chunk_row = chunk_row + fetch_offset + 1;
+		if (chunk_row < -1) {
+			chunk_row = -1;
+		}
+		prior_chunk_row = chunk_row;
+		return SQL_SUCCESS;
+	}
+
+	// FetchOffset < 0    AND | FetchOffset |    > LastResultRow AND |FetchOffset | > RowsetSize
+	if ((fetch_offset < 0) && (std::abs(fetch_offset) > chunk_row) && ((SQLULEN)std::abs(fetch_offset) > rowset_size)) {
+		// Before start, return to BOF
+		return FetchFromBOF();
+	}
+
+	// FetchOffset < 0    AND | FetchOffset |   > LastResultRow AND | FetchOffset | <= RowsetSize
+	if ((fetch_offset < 0) && (std::abs(fetch_offset) > chunk_row) &&
+	    ((SQLULEN)std::abs(fetch_offset) <= rowset_size)) {
+		chunk_row = prior_chunk_row = 0;
+		return SQL_SUCCESS;
+	}
+
+	// FetchOffset = 0, i.e.,return to BOF
+	if (fetch_offset == 0) {
+		if (!chunks.empty()) {
+			return FetchFromBOF();
+		}
+		return FetchNext(stmt);
+	}
+
+	// 1 <= FetchOffset <= LastResultRow
+	if ((fetch_offset >= 0) && (fetch_offset < (SQLLEN)current_chunk->size())) {
+		chunk_row = fetch_offset - 1;
+		prior_chunk_row = chunk_row;
+		return SQL_SUCCESS;
+	}
+
+	// FetchOffset > LastResultRow
+	if (fetch_offset >= (SQLLEN)current_chunk->size()) {
+		return FetchNext(stmt);
+	}
+
+	return SQL_ERROR;
+}
+
+SQLRETURN OdbcFetch::FetchNextChunk(SQLULEN fetch_orientation, OdbcHandleStmt *stmt, SQLLEN fetch_offset) {
+	if (cursor_type == SQL_CURSOR_FORWARD_ONLY || fetch_orientation == SQL_FETCH_NEXT) {
```

Comment:
Shouldn't the cursor instead return an error if the type is set to `SQL_CURSOR_FORWARD_ONLY` and a non-forward fetch is attempted (e.g. `SQL_FETCH_PRIOR`)?

Reply:
Yeah, that makes sense.